### ISSUE TITLE
Issue with CopyFiles module (logic used when checking file hashes)

### ISSUE THREAD

**Issue by elistia** [statiqdev/Statiq.Framework#189]
Hi, I have just started using Statiq so apologies in advance if I am not understanding how this should work...

I am converting a simple set of Markdown files (with images) to html. I started by broadly following this blog post http://www.banna.us/posts/2021-03-06.

Specifically I am using the following process:
```csharp
.BuildPipeline("Copy files", builder => builder
        .WithProcessModules(new CopyFiles("**/*.jpg", "**/*.png", "**/*.css")))
```

This worked fine to copy my image files, but I found that the next time I ran the program my image files were removed from the output. Then the next time they appeared again and this then repeated...

Debug output showed the following when my program ran and the files were already in the output folder:

```
[INFO] Cleaned 2 files not written to output directory c:/dev/test/out during current execution
```

I think the issue is in the `CopyFilesAsync` method in `Statiq.Core/Modules/IO/CopyFiles.cs`

Code is as follows (from around line 138):

```csharp
  // Did we copy this file last time and has no one messed with it?
  int outputFileHash = await outputFile.GetCacheCodeAsync();
  if (context.FileSystem.WriteTracker.TryGetPreviousWrite(outputFile.Path, out int previousWriteHash)
      && previousWriteHash == outputFileHash)
  {
      // We wrote this file last time, it still exists, and it hasn't changed
      // Now check if the content we're about to write is the same as last time
      if (context.FileSystem.WriteTracker.TryGetPreviousContent(outputFile.Path, out int previousContentHash)
          && previousContentHash == outputFileHash)
      {
          // We used the same content last time, so we can skip writing this file
          // Make sure to add the appropriate entries so it looks like we wrote it this time
          context.LogDebug($"Skipped copying file {outputFile.Path.FullPath} from {input.Source.ToSafeDisplayString()} because it already exists and the content is the same");
          context.FileSystem.WriteTracker.TrackWrite(outputFile.Path, previousWriteHash, false);
          context.FileSystem.WriteTracker.TrackContent(outputFile.Path, previousContentHash);
      }
  }
  else
  {
      // Copy to the destination
      await file.CopyToAsync(outputFile, cancellationToken: context.CancellationToken);
      context.FileSystem.WriteTracker.TrackContent(outputFile.Path, outputFileHash); // The file write hash will be tracked when the file write actually occurs
      context.LogDebug("Copied file {0} to {1}", file.Path.FullPath, outputFile.Path.FullPath);
  }
```

The code is generating `outputFileHash` based on the destination file. If the file doesn't exist then the hash is based on the file name. If the file does exist then the hash is based on the filename, length, creation time and last write time.

This hash is then compared to two other hashes (the Write Hash and the Content Hash) which are stored in a cache file. 

**Case 1.**
When the output file has not been copied the previous time and is therefore not in the cache, the outputFileHash will be based on the filename.

The code will go directly to the section labelled `//Copy to the destination` as the previousWriteHash will be zero. The hash that has been generated (just from the filename) will be stored in the cache as the 'Content Hash'.
The file will then be written and a 'Write Hash' added to the cache (calculated somewhere else and based on the filename, length, creation time and last write time presumably?

**Case 2.**
The next time the code is run, the file exists in the output location so the outputFileHash will be calculated on the filename, length, creation time and last write time of this file.

The following code will evaluate to true, as the previousWriteHash was also calculated on the filename, length, creation time and last write time of this file (assuming the file has not been tampered with).

```csharp
if (context.FileSystem.WriteTracker.TryGetPreviousWrite(outputFile.Path, out int previousWriteHash)
                    && previousWriteHash == outputFileHash)
```

However, the next if statement will always evaluate to false. 

```csharp
    // We wrote this file last time, it still exists, and it hasn't changed
    // Now check if the content we're about to write is the same as last time
    if (context.FileSystem.WriteTracker.TryGetPreviousContent(outputFile.Path, out int previousContentHash)
        && previousContentHash == outputFileHash)
```

We are still using the same outputFileHash, but this time comparing it to the previous**Content**Hash which was generated just from the filename.

Therefore the code drops out of this if statement with no action. The file is not copied, but neither is it 'Skipped' and the entries added to the tracker to make it look as though it has been added.
Therefore, somewhere later in the process (?FileCleaner.cs) the file is being deleted as it exists but hasn't been added as part of this process.

_Apologies for this being incredibly long winded, but mainly for my benefit as I try and understand what is going on :-)_

In the short term this could be fixed by removing the following code:

```diff
// We wrote this file last time, it still exists, and it hasn't changed
- // Now check if the content we're about to write is the same as last time
- if (context.FileSystem.WriteTracker.TryGetPreviousContent(outputFile.Path, out int previousContentHash)
-     && previousContentHash == outputFileHash)
- {
    // We used the same content last time, so we can skip writing this file
    // Make sure to add the appropriate entries so it looks like we wrote it this time
    context.LogDebug($"Skipped copying file {outputFile.Path.FullPath} from {input.Source.ToSafeDisplayString()} because it already exists and the content is the same");
    context.FileSystem.WriteTracker.TrackWrite(outputFile.Path, previousWriteHash, false);
    context.FileSystem.WriteTracker.TrackContent(outputFile.Path, previousContentHash);
- }
```

I am very happy to do a pull request for the above change if that would be helpful?

I'm not sure I understand the use of the Content Hash here. The Content Hash seems not to be based on the content of the file at all, just on the name. I am therefore not sure what purpose it serves? Was it intended that the Content Hash be calculated in a different way for these files?

Many thanks for reading this far. Statiq is great and very much looking forward to looking at Docs (it is project documentation that I am using this for) :-)

**Comment by daveaglick** [comment]
Fantastic analysis, thanks! Made it really easy to track down the underlying problem. For background, I only very recently added this whole caching mechanism that looks at files in the output folder and tries to avoid disk writes when possible (since disk I/O is really expensive).

You correctly identified that there are two actual caches at play - the reason we need both a "content" cache and a "write" cache is because the former caches the source data for a potential write and the latter caches what we _think_ we wrote to disk (though we also need to account for someone/something changing what's on disk out from under us). In situations like Markdown or Razor rendering the content cache is actually based on in-memory content as the result of the rendering operation, but for plain file copies we don't want to read the whole file (that defeats the purpose of avoiding disk I/O) so the "content" cache is based on file metadata under the assumption that if that changes, the file contents probably changed too.

Which brings us to the bug! We need to store and check both the content hash and the write hash, but the bug is that the `CopyFiles` module is using the same values for both (which is where I think you got thrown off too). This didn't show up in my testing because I rarely use `CopyFiles` - the pipelines built into Statiq Web actually do a `ReadFiles` followed by a `WriteFiles` which behaves a little differently since the file ends up in an `IDocument` as it flows from the read module to the write module. However, `CopyFiles` is absolutely valuable for custom pipelines like this so I'm glad to found the bug!

I think I've got a fix ready, just doing some final testing. I found a couple other little nits while I was in there too, so all-in-all a very productive bug to fix. I'll get a new release of Statiq Framework out once it's confirmed (though I probably won't cascade this one update through to a new Statiq Web release just yet).
